### ISSUE TITLE
x/build: use LLD linker on linux-amd64-clang builder?

### ISSUE THREAD

**Issue by cherrymui** [golang/go#59482]
The LLD linker is widely used by users. Occasionally there are some subtle differences in behavior between the C linkers (BFD, Gold, LLD), that cause issues for us (e.g. we may have assumptions based on the BFD linker's specific behavior (e.g. #59466), or we may uncover a bug in the C linker itself.)

As far as I can tell, we don't have much LLD coverage on the builders. I think it is probably good to have one. The `linux-amd64-clang` builder already uses LLVM based C compiler. So it may be a good fit. (gcc+lld is possible but probably not a popular configuration.)

Thoughts? Thanks.

cc @prattmic (was chatting with Michael about this the other day)

**Comment by prattmic** [comment]
I agree we should have a clang+lld builder. My question is whether we should also keep a builder with the existing configuration (is it clang+bfd or clang+gold?).

**Comment by mknyszek** [comment]
I'm going to guess that we should probably have builders for every popular C linker (or whatever we're willing to say we explicitly support).

Logistical question: what can this builder share with other builders? Could it, for example, share `./make.bash` output?

**Comment by cherrymui** [comment]
Now we build the toolchain without cgo, right? If so, the result of `make.bash` does not depend on the C linker at all. So we can share it.

Even if it still uses cgo, we probably can share. We just don't use LLD to link some tool binaries (the go command, the pprof command?), and that is probably fine, and we'll probably have the same coverage when linking test binaries.

**Comment by cherrymui** [comment]
We could have builder for clang+bfd, if the builder resource is fine. I think that is probably also a popular configuration that users use.

**Comment by motiejus** [comment]
For the record (this was mentioned in the previous issue, but worth repeating here) this exact combination would be very welcome for those of us who compile C/C++ files using `zig cc`, as Zig is actually clang+lld behind the scenes.

**Comment by motiejus** [comment]
Related: https://github.com/golang/go/issues/59666

**Comment by dmitshur** [comment]
CC @golang/release.

**Comment by And-ZJ** [comment]
Hello, I'm sorry for the sudden interruption.
I noticed that the gold linker may have been deprecated, see [GNU Binutils 2.44 Released
](https://lists.gnu.org/archive/html/info-gnu/2025-02/msg00001.html) .
I think it's necessary to introduce the `lld` linker for testing, so that we can find out if there's any behavior that depends on a particular linker. ( I found [#69466](https://github.com/golang/go/issues/69466) )
I noticed that under `linux/arm64`, the `gold` linker is used by default for the dynlinking scenario. Maybe this will need to be replaced with other linkers in the future.
Thank you!

**Comment by cherrymui** [comment]
I think it is a good idea to test lld (that's why I filed the issue).

In the Go linker, we probably can switch off from defaulting to gold.
